api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto

```diff
@@ -269,4 +270,12 @@ message ExtProcOverrides {
 
   // Set a different gRPC service for this route than the default.
   config.core.v3.GrpcService grpc_service = 5;
+
+  // Additional metadata to include in streams initiated to the GrpcService.
+  // This can be used for scenarios in which additional ad hoc authorization
+  // headers (e.g. ``x-foo-bar: baz-key``) are to be injected or when a route
+  // needs to partially override inherited metadata. For more information,
+  // including details on header value syntax, see the documentation on
+  // :ref:`custom request headers <config_http_conn_man_headers_custom_request_headers>`.
+  repeated config.core.v3.HeaderValue metadata = 6;
```

Comment:
It is unclear to me which field this is overriding?
Also, are these essentially request-headers that need to be added?

Reply:
for more info check the example https://github.com/envoyproxy/envoy/files/13885400/ext_proc_metadata_per_route_example.txt

Reply:
Thanks for explaining this.
The `ExtProcOverrides` type allows overriding fields that are in `ExternalProcessor` for a specific route. For example, if the route needs a different processing_mode, the a value in [ExtProcOverrides->processing_mode](https://github.com/envoyproxy/envoy/blob/143ef4a82d26a75ccd460563f5e48044607883c5/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L255) will override the value from the [ExternalProcesseor->processing_mode](https://github.com/envoyproxy/envoy/blob/143ef4a82d26a75ccd460563f5e48044607883c5/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L119).

IIUC the `initial_metadata` field you are referring to is inside the [GrpcService type](https://github.com/envoyproxy/envoy/blob/234dcc646cbfc1cd512742fc92ad704f514b5948/api/envoy/config/core/v3/grpc_service.proto#L302). If that's the case, why not override the [grpc_service](https://github.com/envoyproxy/envoy/blob/143ef4a82d26a75ccd460563f5e48044607883c5/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L272) field?

Reply:
> IIUC the initial_metadata field you are referring to is inside the [GrpcService type](https://github.com/envoyproxy/envoy/blob/234dcc646cbfc1cd512742fc92ad704f514b5948/api/envoy/config/core/v3/grpc_service.proto#L302). If that's the case, why not override the [grpc_service](https://github.com/envoyproxy/envoy/blob/143ef4a82d26a75ccd460563f5e48044607883c5/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L272) field?

@adisuissa it is because when you do provide `ExtProcPerRoute.overrides.grpc_service` [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#extensions-filters-http-ext-proc-v3-extprocperroute)] you 

1. can't provide only  `ExtProcPerRoute.overrides.grpc_service.initial_metadata` - it is necessary to provide at least one of `grpc_service.envoy_grpc` or `grpc_service.google_grpc`.
2. even if we could do (1) the original semantics of `ExtProcPerRoute.overrides.grpc_service` is that it is completely REPLACES the original inherited configuration from the ExternalProcessor [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#extensions-filters-http-ext-proc-v3-externalprocessor)] whilst we need to only extend and maybe override - basically inherit the parents configuration with the small changes.
3. even if (2) was possible to do with just `ExtProcPerRoute.overrides.grpc_service` without the issue of (1) the `grpc_service.initial_metadata` name suggests that it does not have the extending semantics but the replaсement. 
4. then GrpcService [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/grpc_service.proto#config-core-v3-grpcservice)] is the global configuration component reused in many places of configuration so we can't change its semantics. as well as no one would appreciate the change in the semantics of `ExtProcPerRoute.overrides.grpc_service.initial_metadata` if we had tried to modify it.

We needed the option to extend (A) and modify (B) inherited (C) metadata configuration for _ext_proc_ service on the per route basis WITHOUT overriding the GrpcService [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/grpc_service.proto#config-core-v3-grpcservice)] configuration itself - that is why we extended the `ExtProcPerRoute.overrides` with the `ExtProcPerRoute.overrides.metadata` repeatable field. It was impossible to do before this patch.

With this patch we now can have a big parent ExternalProcessor [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#extensions-filters-http-ext-proc-v3-externalprocessor)] configuration with proper GrpcService [[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/grpc_service.proto#config-core-v3-grpcservice)] and corresponding `initial_metadata` and only extend it in our thousands of routes.

Reply:
The `ExtProcOverride` semantics is always replacement.
From an API perspective, I'm concerned about the case where both `grpc_service` and `metadata` are overridden. In this case, it is not clear what the `initial_metadata` should be.

Is (1) you stated above the main issue you see in using the current `grpc_service` override option?

I think another option could be adding a new type that customizes the GrpcService, where other features can be added (not only overriding the `initial_metadata`, but other fields as needed). @envoyproxy/api-shepherds may have a different perspective on this.

Reply:
> The ExtProcOverride semantics is always replacement.

In this patch, we allowed for whatever `initial_metadata` configuration there is to converge first (merge parent ExternalProcessor with ExtProcPerRoute that actually replaces the `grpc_service` configuration completely with the latest variant), and only then be extended with the `ExtProcOverride.metadata`.

> From an API perspective, I'm concerned about the case where both grpc_service and metadata are overridden. In this case, it is not clear what the initial_metadata should be.

This has been resolved surprisingly well. See the example [ext_proc_metadata_per_route_example.yaml](https://github.com/envoyproxy/envoy/files/13885400/ext_proc_metadata_per_route_example.txt) the part with `key3` (`CLIENT_PROXY=http://127.0.0.1:9911 ./client.py set key3 1`):

When you have this:

```
              - match:
                  prefix: "/kv.KV"
                  headers:
                  - name: "key"
                    exact_match: "key3"
                route:
                  cluster: backend-proxy
                typed_per_filter_config:
                  envoy.filters.http.ext_proc:
                    "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExtProcPerRoute
                    overrides:
                      grpc_service:
                        envoy_grpc:
                          cluster_name: ext-proc-proxy
                        timeout: 30s
                        initial_metadata:
                          - key: "c"
                            value: "c"
                          - key: "d"
                            value: "d"
                      metadata:
                        - key: "d"
                          value: "blah"
                        - key: "g"
                          value: "blah"
          http_filters:
          - name: envoy.filters.http.ext_proc
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExternalProcessor
              failure_mode_allow: false
              async_mode: false
              allow_mode_override: true
              message_timeout: 360s
              grpc_service:
                envoy_grpc:
                  cluster_name: ext-proc-proxy
                timeout: 60s
                initial_metadata:
                  - key: "a"
                    value: "a"
                  - key: "b"
                    value: "b"
```

you will get

```
    [debug][router] [source/common/router/router.cc:731] [Tags: "ConnectionId":"0","StreamId":"5894068517272876876"] router decoding headers:
    ':method', 'POST'
    ':path', '/envoy.service.ext_proc.v3.ExternalProcessor/Process'
    ':authority', 'ext-proc-proxy'
    ':scheme', 'http'
    'te', 'trailers'
    'content-type', 'application/grpc'
    'c', 'c'
    'd', 'blah'
    'g', 'blah'
    'x-envoy-internal', 'true'
    'x-forwarded-for', '172...'
```

The `ExtProcPerRoute.overrides.grpc_service.initial_metadata` replaces the original `ExternalProcessor.grpc_service.initial_metadata` and then is perfectly extended by the `ExtProcOverride.metadata`.

> Is (1) you stated above the main issue you see in using the current grpc_service override option?

Mostly because of (2). `ExtProcOverride.grpc_service` completely replaces parent `grpc_service`.

Reply:
I am wondering if we should name it `grpc_initial_metadata` explicitly? So that it is clear that this field is used for partial grpc service override, although it is clearly described in comment.
I don't feel strongly and will let @adisuissa @yanavlasov  and other folks make the call.

Reply:
The issue with merge-semantics is that it hinders the understandability of the config.
The filter-config override can appear in multiple-levels: Virtual-Host, Route, and Cluster-Weight, which make the merge semantics unclear.
cc @envoyproxy/api-shepherds for additional views.

Reply:
> I am wondering if we should name it `grpc_initial_metadata` explicitly? So that it is clear that this field is used for partial grpc service override, although it is already described in comment. I don't feel strongly and will let @adisuissa @yanavlasov and other folks make the call.

If the API field remains then yeah the field should be exact as possible so one can understand exactly which metadata field is set by it.

Reply:
as we have https://github.com/envoyproxy/envoy/pull/30747 merged first with `metadata_options = 6;` I have renamed this PR field into `repeated config.core.v3.HeaderValue grpc_metadata = 7;`. Now it must be much more clear what metadata it is capable of extending.

Reply:
we now have it as `grpc_initial_metadata`